src/Router/index.tsx

```diff
@@ -1,123 +1,67 @@
-    location: '',
-    params: {},
-    context: {},
-    routes: props.routes,
-    routedElement: undefined,
-    forceRefresh: 0
-  };
+const browserHistory = createBrowserHistory()
 
-  const [state, setState] = useState(initialState);
+const getComponentFromRoute = (route: Route, router: IRouter) => {
+  let guardReturn: React.ReactNode = undefined;
 
-  const setLocation = (location: string) => {
-    if (state.location !== location) {
-      history.current?.push(location);
-    } else {
-      setState({
-        ...state,
-        forceRefresh: state.forceRefresh + 1
-      });
+  if (route.guards) {
+    let nexted = false;
+    for (const guard of route.guards) {
+      guardReturn = guard(router, () => {
+        nexted = true
+        return undefined
+      })
+      if (!nexted && guardReturn) break
     }
-  };
-
-  const setContext = (context: Object) => {
-    setState({
-      ...state,
-      context: Object.assign(state.context, context)
-    });
-  };
+  }
 
-  useEffect(() => {
-    history.current = createBrowserHistory();
-    setState({
-      ...state,
-      location: history.current.location.pathname
-    });
-
-    const unlisten = history.current.listen(update => {
-      setState({
-        ...state,
-        location: update.location.pathname
-      });
-    });
+  return guardReturn ?? route.component
+}
 
-    return () => {
-      unlisten();
-    };
-  }, []);
+export const Router = ({ routes, children }: IRouterProps) => {
+  const [location, setLocation] = useState<string>(browserHistory.location.pathname)
+  const [context, setContext] = useState<object>({})
 
+  // Listen to location changes
   useEffect(() => {
-    let route = state.routes.find(route => {
-      return Path.match(route.path, state.location);
+    browserHistory.listen(update => {
+      setLocation(update.location.pathname);
     });
+  }, []);
 
-    if (route) {
-      let guardReturn: React.ReactNode = undefined;
-      let nexted = false;
-
-      if (route.guards) {
-        for (const guard of route.guards) {
-          guardReturn = guard({
-            location: state.location,
-            setLocation,
-            setContext,
-            context: state.context,
-            params: state.params,
-          }, () => {
-            nexted = true
-            return undefined
-          });
-          if (!nexted && guardReturn) break;
-        }
-      }
-
-      setState({
-        ...state,
-        params: Path.parse(route.path, state.location),
-        routedElement: guardReturn ?? route.component
-      });
-
-    }
-
+  const route = routes.find(route => Path.match(route.path, location));
 
+  if (!route) {
+    console.error(`Current location ${location} did not match any defined route`)
+    return null;
+  }
 
-  }, [state.location, state.forceRefresh]);
+  const params = Path.parse(route.path, location)
+  const component = getComponentFromRoute(route, { location, setLocation, context, setContext, params })
```

Comment:
Both this and the route computation could be memoized but I think every re-render will need to recompute it anyway so I think it's safe to just roll with it.

On very heavy routings (hundreds of thousands of routes times hundreds of thousands of guards per route, this could be slow) but if the case ever comes up we could just optimize it.